src/types/keys.ts

```diff
@@ -20,34 +22,43 @@ export interface BaseKey {
  * hatch to support any exotic key type in the future.
  */
 export interface Key extends BaseKey {
+  id: string;
+  privateKey: string;
+}
+
+export interface NewKey extends BaseKey {
```

Comment:
I'm not sure what "NewKey" means. I think this name should be more specific about how this differs from Key.

Reply:
The `NewKey` type doesn't have the `id`. I tried making it optional in `BaseKey`, but that would require adding guards to several points of the code (otherwise Typescript is gonna complain about it).

```
src/fixtures/keys.ts:37:3 - error TS2322: Type '{ encrypterName: string; salt: string; encryptedBlob: string; creationTime?: number | undefined; extra?: any; id?: string | undefined; modifiedTime?: number | undefined; path?: string | undefined; publicKey: string; type: string; }' is not assignable to type 'EncryptedKey'.
  Types of property 'id' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.

 37   return {
      ~~~~~~~~
 38     ...key,
    ~~~~~~~~~~~
... 
 41     encryptedBlob: `${privateKey}password`,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 42   };
    ~~~~

src/KeyManager.ts:132:27 - error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

132     this._writeIndexCache(newKey.id, newKey);
                              ~~~~~~~~~

src/KeyManager.ts:297:31 - error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

297         this._writeIndexCache(decryptedKey.id, decryptedKey);
                                  ~~~~~~~~~~~~~~~

src/plugins/IdentityEncrypter.ts:10:3 - error TS2322: Type '({ key }: { key: Key; }) => Promise<{ encryptedBlob: string; encrypterName: string; salt: string; creationTime?: number | undefined; id?: string | undefined; modifiedTime?: number | undefined; }>' is not assignable to type '(params: EncryptParams) => Promise<EncryptedKey>'.
  Type 'Promise<{ encryptedBlob: string; encrypterName: string; salt: string; creationTime?: number | undefined; id?: string | undefined; modifiedTime?: number | undefined; }>' is not assignable to type 'Promise<EncryptedKey>'.
    Type '{ encryptedBlob: string; encrypterName: string; salt: string; creationTime?: number | undefined; id?: string | undefined; modifiedTime?: number | undefined; }' is not assignable to type 'EncryptedKey'.
      Types of property 'id' are incompatible.
        Type 'string | undefined' is not assignable to type 'string'.
          Type 'undefined' is not assignable to type 'string'.

10   encryptKey({ key }: { key: Key }) {
     ~~~~~~~~~~

  src/types/keys.ts:81:3
    81   encryptKey(params: EncryptParams): Promise<EncryptedKey>;
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The expected type comes from property 'encryptKey' which is declared here on type 'Encrypter'

src/plugins/ScryptEncrypter.ts:25:5 - error TS2322: Type '{ encryptedBlob: string; encrypterName: string; salt: string; creationTime?: number | undefined; id?: string | undefined; modifiedTime?: number | undefined; }' is not assignable to type 'EncryptedKey'.
  Types of property 'id' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.

 25     return {
        ~~~~~~~~
 26       ...props,
    ~~~~~~~~~~~~~~~
... 
 29       salt,
    ~~~~~~~~~~~
 30     };
    ~~~~~~


Found 5 errors.

```

To be honest, I'm not sure what's the best approach, so I'm open to any suggestions.

Reply:
I think having the type is fine, i just think it’s name isn’t descriptive enough. You could name it IDlessKey, but that doesn’t explain when or why you’d need that type instead of Key. So I guess what I’m asking is, when would someone need to provide an id-less key instead of a regular one?